python/lsst/daf/butler/dimensions/_record_table.py

```diff
@@ -0,0 +1,232 @@
+            return result
+        else:
+            return self._get_record_at(self._table, index)
+
+    def extend(self, records: Iterable[DimensionRecord]) -> None:
+        """Add new rows to the end of the table.
+
+        Parameters
+        ----------
+        records : `~collections.abc.Iterable` [ `DimensionRecord` ]
+            Dimension records to add to the table.
+        """
+        batches: list[pa.RecordBatch] = self._table.to_batches()
+        batches.append(self._make_batch(records, self._table.schema))
+        self._table = pa.Table.from_batches(batches, self._table.schema)
+
+    def column(self, name: str) -> pa.ChunkedArray:
+        """Return a single column from the table as an array.
+
+        Parameters
+        ----------
+        name : `str`
+            Name of the column.  Valid options are given by
+            `DimensionElement.schema.names`, and are the same as the attributes
+            of the dimension records.
+
+        Returns
+        -------
+        array : `pyarrow.ChunkedArray`
+            An array view of the column.
+        """
+        return self._table.column(name)
+
+    def to_arrow(self) -> pa.Table:
+        """Return a Arrow table holding the same records."""
+        return self._table
+
+    def _make_batch(self, records: Iterable[DimensionRecord], arrow_schema: pa.Schema) -> pa.RecordBatch:
+        """Make a `pyarrow.RecordBatch` from an iterable of `DimensionRecord`.
+
+        Parameters
+        ----------
+        records : `~collections.abc.Iterable` [ `DimensionRecord` ]
+            Records to add.
+        arrow_schema : `pyarrow.Schema`
+            Arrow schema for the record batch.
+
+        Returns
+        -------
+        batch : `pyarrow.RecordBatch`
+            Record batch holding the records.
+        """
+        list_columns: list[list[Any]] = [list() for _ in self._converters]
+        for record in records:
+            for converter, column in zip(self._converters, list_columns):
+                converter.append(getattr(record, converter.name), column)
+        array_columns = [
+            converter.finish(column) for converter, column in zip(self._converters, list_columns)
+        ]
+        return pa.record_batch(array_columns, arrow_schema)
+
+    def _get_record_at(self, table: pa.Table | pa.RecordBatch, index: int) -> DimensionRecord:
+        """Construct a `DimensionRecord` from a row in the table.
+
+        Parameters
+        ----------
+        table : `pyarrow.Table` or `pyarrow.RecordBatch`
+            Table or record batch to get values from.
+        index : `int`
+            Index of the row to extract.
+
+        Returns
+        -------
+        record : `DimensionRecord`
+            Dimension record representing a table row.
+        """
+        return self._element.RecordClass(
+            **{k: table.column(j)[index].as_py() for j, k in enumerate(self._element.schema.all.names)}
```

Comment:
Not at all necessary right now, but there might be a possibility for future optimization by lazily converting the data from the columns as they are used instead of immediately converting the entire row.

Reply:
In this context I'm more concerned with being not holding on to the row-based inputs so they can be garbage collected.  But your instincts are good: there are clear analogs elsewhere in the codebase of this point, in terms of us doing work on `DimensionRecord` fields that are never accessed at all, and I've been mulling over possible changes to the `DataCoordinate` public interface to maybe help with that, since the alternative is a lot of lazy-loading that extends through many, many layers.